### PR TITLE
Reapply "Enable VA-API support in FreeRDP"

### PR DESCRIPTION
**Pull request by travier**
Reapply "Enable VA-API support in FreeRDP"

This reverts commit c70cec9d5a8d929a74f3820526ac602d0b3a007f.

---

See: https://github.com/flathub/org.kde.krdc/issues/81

### CODE REVIEW

**Comment by JoelleJS** [comment]
As mentioned in #81 this works on my PC (AMD GPU) but I'd like to test this one on two laptops of mine with other GPU brands/Linux distributions.

**Comment by travier** [comment]
Thanks for testing! I'll wait until you get back to us with results.

**Comment by JoelleJS** [comment]
I couldn't get VA-API working on NVIDIA on Linux Mint for the life of me (in general, not just in KRDC). So I only tested two laptops with Intel graphics. It works well.

The issue with out of bounds resolutions remains. As a workaround you can force software decoding by setting a bogus value for `LIBVA_DRIVER_NAME`, e.g. `--env="LIBVA_DRIVER_NAME=asdfg"`.

Not ideal but perhaps acceptable given the benefits of hardware acceleration.

**Comment by twelho** [comment]
I've managed to test this with an Intel-based laptop connecting to a Windows 10 + NVIDIA setup. The video decoder does kick in, but there is some kind of major memory corruption happening, and view is basically completely garbled, flickering and overlays tiles from other applications. Haven't seen this with other applications using hardware decoding. Selecting "Disable All Acceleration" in the connection settings does however successfully disable hardware decoding and the picture is clean again. Maybe this could also work around the out of bounds resolution @JoelleJS ?

I also had version 24.08.3 with https://github.com/flathub/org.kde.krdc/pull/42 still kicking around, same issues with that... Maybe I need to re-boot and check again. Or it could also be an issue of the RDP host.

**Comment by JoelleJS** [comment]
@twelho Using `Disable All Acceleration` makes the program not respond for a while and then output
```
[14:09:11:594] [2:2] [ERROR][com.freerdp.core.connection] - Timeout waiting for activation
KRDC: Unable to connect
kf.notifications: Failed to play sound with canberra: File or data not found
[14:09:13:327] [2:2] [ERROR][com.freerdp.core] - freerdp_abort_connect:freerdp_set_last_error_ex ERRCONNECT_CONNECT_CANCELLED [0x0002000B]
```
The remote is an Arch Linux machine running KRDP integrated in Plasma.

**Comment by twelho** [comment]
A reboot didn't fix it in my case, the picture still becomes brighter and brighter with hardware acceleration until it's completely white. In the debug log output I spotted the following lines:

```
[14:58:20:522] [2:2] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_RGBX32
[14:58:20:522] [2:2] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_BGRA32
```

So it might be, that the framebuffer format conversions are failing. Nevertheless, disabling hardware acceleration does fix it, so not a deal breaker in this case. No idea why it doesn't work without hwaccel for you @JoelleJS , seems quite strange that that would have an effect on the connetion. What is the RDP server exactly? AFAIK KDE's integrated tool for that is Krfb, but it doesn't support RDP, only VNC (at least on my machine).

**Comment by ctrlaltca** [comment]
> the picture still becomes brighter and brighter with hardware acceleration until it's completely white.

@twelho are you using wayland? This is a known problem of krdc under wayland when using the freerdp2 backend.
What's happening is that the alpha channel in the image is wrong and each transparent image is getting painted over the previous ones, ending up in a complete white image.

Freerdp3 support has been added in krdc's master and it fixes this bug; it's targeted for the next version (25.04 i guess).
Unfortunately by now the only workaround I know is to disable h264 in the connection and use eg. RemoteFX instead.

**Comment by twelho** [comment]
> @twelho It wasn't properly announced but the remote desktop functionality in the Plasma system settings is backed by KRDP since Plasma 6.1.

Hmm, so I'm running 6.2.5 right now but still didn't see the option, and I just now realized that I installed `krdc` instead of `krdp` (Fedora Kinoite has neither installed by default). Let me check again tonight when I can restart my session.

> @twelho are you using wayland? This is a known problem of krdc under wayland when using the freerdp2 backend. What's happening is that the alpha channel in the image is wrong and each transparent image is getting painted over the previous ones, ending up in a complete white image.
> 
> Freerdp3 support has been added in krdc's master and it fixes this bug; it's targeted for the next version (25.04 i guess). Unfortunately by now the only workaround I know is to disable h264 in the connection and use eg. RemoteFX instead.

Yes, I'm indeed using Wayland. This sounds exactly like what I'm experiencing. I'll also try building this Flatpak from https://invent.kde.org/network/krdc/-/merge_requests/151 then with this patch tonight, let's see if I can get clear hardware-accelerated output.

**Comment by twelho** [comment]
Sorry for the delay, finally got back to this. We're still waiting for KRDC [v25.04](https://invent.kde.org/network/krdc/-/network/master?ref_type=heads) with FreeRDP 3 support to get tagged to unblock this. However, I did now test this patch against v24.12.3, that is, the current version still based on FreeRDP 2. While a Windows RDP server will still corrupt the output, I did get the newly built-in KDRP server in Plasma working without corruption, both with an without hardware acceleration, as confirmed by `intel_gpu_top`.

TL;DR: Still waiting for a release in the KRDC v25 branch with FreeRDP 3 support before moving forward.

**Comment by twelho** [comment]
Now that KRDC 25.04 has made its way to this Flatpak I attempted this patch again with much better success. Connecting to a Plasma 6 machine over the new built-in RDP server with `-DWITH_VAAPI:BOOL=ON` works, and I can see a tiny sliver of video acceleration usage in `intel_gpu_top`. Connecting to Windows also works without corruption now and the logs state that hardware acceleration is in use, but it's either only very lightly used or can't actually be used due to the weird pixmap formats. Regardless, the default cases with VA-API enabled don't cause issues anymore (at least on my end), so I'd say it's safe to enable by default again.

**Comment by twelho** [comment]
> 🚧 [Test build succeeded](https://github.com/flathub-infra/vorarbeiter/actions/runs/15413791534). To test this build, install it from the testing repository:
> 
> ```
> flatpak install --user https://dl.flathub.org/build-repo/189438/org.kde.krdc.flatpakref
> ```

Tested this now, still works for me. Connecting to a Plasma 6 session with multiple monitors attached works as well and hardware acceleration is being utilized. There's quite a lot of latency (1-2 seconds), but that occurs with both the latest Flathub build as well as this build, so I don't think VA-API is causing it.
